plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/DebugService.java

```diff
@@ -0,0 +1,216 @@
+ */
+@JsonSegment("debug")
+@ImplementedBy(DebugServiceNullImpl.class)
+public interface DebugService {
+
+	/** Sets the client to communicate to */
+	void connect(LanguageClient client);
+
+	/** Sets the level of verbosity of Log4j. @see {@link LogLevel} */
+	@JsonRequest
+	CompletableFuture<Void> setVerboseLevel(String level);
+
+	/** Prints debug information on the output channel */
+	@JsonRequest
+	CompletableFuture<Void> printDebugInfo();
+
+	/** Default null implementation */
+	class DebugServiceNullImpl implements DebugService {
+		@Override
+		public void connect(LanguageClient client) {
+			// nothing to do
+		}
+
+		@Override
+		public CompletableFuture<Void> setVerboseLevel(String level) {
+			return CompletableFuture.failedFuture(new UnsupportedOperationException());
+		}
+
+		@Override
+		public CompletableFuture<Void> printDebugInfo() {
+			return CompletableFuture.failedFuture(new UnsupportedOperationException());
+		}
+	}
+
+	/** Default implementation */
+	class DebugServiceDefaultImpl implements DebugService {
+		private static final Logger LOG = LogManager.getLogger(DebugServiceDefaultImpl.class);
+
+		@Inject
+		private QueuedExecutorService executerService;
+
+		@Inject
+		private ConcurrentIndex fullIndex;
+
+		private LanguageClient client;
+
+		@Override
+		public void connect(LanguageClient _client) {
+			this.client = _client;
+		}
+
+		@Override
+		public CompletableFuture<Void> setVerboseLevel(String level) {
+			Level actualLevel = Level.toLevel(level.toUpperCase());
+			Logger.getRootLogger().setLevel(actualLevel);
+			LOG.info("log level set to " + Logger.getRootLogger().getLevel());
+			return CompletableFuture.completedFuture(null);
+		}
+
+		@Override
+		public CompletableFuture<Void> printDebugInfo() {
+			String msgText = getDebugInfo();
+			MessageParams message = new MessageParams();
+			message.setMessage(msgText);
+			message.setType(MessageType.Log);
+			client.logMessage(message);
+			return CompletableFuture.completedFuture(null);
+		}
+
+		private String getDebugInfo() {
+			String info = "==   DEBUG INFO   ==\n";
+			info += getThreadDump();
+			info += "\n--------------\n";
+			info += getExecuterServiceDump();
+			info += "\n--------------\n";
+			info += getMemoryDump();
+			info += "\n--------------\n";
+			info += getWorkspaceConfigDump();
+			info += "\n== DEBUG INFO END ==\n";
+			return info;
+		}
+
+		// copied and modified from org.eclipse.osgi.framework.util.ThreadInfoReport
+		private static String getThreadDump() {
+			long currentId = Thread.currentThread().getId();
+			ThreadMXBean threadMXBean = ManagementFactory.getThreadMXBean();
+			StringBuilder dump = new StringBuilder("Thread dump:\n");
+			dump.append("Current ThreadId: ").append(currentId);
+
+			ThreadInfo[] infos = threadMXBean.dumpAllThreads(threadMXBean.isObjectMonitorUsageSupported(),
+					threadMXBean.isSynchronizerUsageSupported());
+
+			for (ThreadInfo info : infos) {
```

Comment:
ThreadInfo.toString() doesn't look to bad either. Should we try to simplify this?